File: packages/sdk/src/domains/sync.ts
/**
 * The sync domain: fetch conversation history the client is missing.
 *
 * A client holds a highest-contiguous sequence number per conversation. When it reconnects, or when a
 * live {@link MessageEvent} arrives with a `seq` past that watermark (a gap), it asks the server for
 * the range in between. The returned {@link MessageEvent}s are still sealed — sync moves ciphertext,
 * not plaintext, because the server has no plaintext to move — so the caller replays each through
 * {@link MessagingDomain.ingest}, which decrypts and routes it exactly as a live delivery.
 *
 * # Why replay order matters
 *
 * History is returned in ascending sequence order, and a client must ingest it in that order. A
 * sender-key distribution ({@link MessageKind.KeyExchange}) has a lower seq than the content it
 * unlocks, so replaying forward hands the messaging layer the key before the messages that need it. A
 * ranged or backwards fetch can still deliver content before its key; the messaging layer's bounded
 * per-sender buffer absorbs that and drains when the distribution replays.
 *
 * # The truncation boundary
 *
 * The server may cap how far back it will serve. A {@link SyncStatus.Truncated} response means older
 * history exists but was not returned, so the UI should render a "messages before here are
 * unavailable" boundary rather than silently presenting a gap as continuity.
 */

import type { Id } from '@migo/wire';
import { OP, encodeSyncRequest, decodeSyncResponse } from '@migo/protocol';
import type { SyncRequest, SyncResponse } from '@migo/protocol';

import type { Rpc } from './rpc.js';

/** Optional bounds on a sync fetch beyond "everything after `haveSeq`". */
export interface SyncOptions {
  /** Fetch up to this sequence number, to fill a detected gap rather than tail the latest. */
  toSeq?: number;
  /** Page older history below `haveSeq` instead of newer history above it. */
  backwards?: boolean;
}

/**
 * Fetch missing conversation history.
 *
 * One instance per client. Holds no watermark itself — the caller tracks the highest contiguous seq
 * it has and passes it as `haveSeq` — so this domain stays a stateless request wrapper.
 */
export class SyncDomain {
  readonly #rpc: Rpc;

  constructor(rpc: Rpc) {
    this.#rpc = rpc;
  }

  /**
   * Fetches up to `limit` messages for a conversation.
   *
   * `haveSeq` is the highest contiguous sequence the caller already holds; the server returns messages
   * after it (or, with {@link SyncOptions.backwards}, before it). The caller replays each message in
   * {@link SyncResponse.messages} through {@link MessagingDomain.ingest} in order, then advances its
   * watermark to {@link SyncResponse.toSeq}. A {@link SyncResponse.more} of `true` means another page
   * is available with the returned `toSeq` as the next `haveSeq`.
   */
  async fetch(
    conversationId: Id,
    haveSeq: number,
    limit: number,
    options: SyncOptions = {},
  ): Promise<SyncResponse> {
    const request: SyncRequest = { conversationId, haveSeq, limit };
    if (options.toSeq !== undefined) {
      request.toSeq = options.toSeq;
    }
    if (options.backwards !== undefined) {
      request.backwards = options.backwards;
    }
    return this.#rpc.call(OP.SYNC, encodeSyncRequest, decodeSyncResponse, request);
  }
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Sync.kt
package com.migo.core.domain

import com.migo.core.protocol.Op
import com.migo.core.protocol.SyncRequest
import com.migo.core.protocol.SyncResponse
import com.migo.core.wire.Id

/**
 * Fetching conversation history this client is missing.
 *
 * A port of `packages/sdk/src/domains/sync.ts`. A client holds the highest contiguous sequence number
 * it has per conversation; when it reconnects, or when a live event arrives with a sequence past that
 * watermark, it asks for the range in between.
 *
 * # Sync moves ciphertext
 *
 * What comes back is a list of sealed `MessageEvent`s -- the same shape a live delivery has, still
 * encrypted -- because the server has no plaintext to return. The caller replays each through
 * [MessagingDomain.ingest], which opens and routes it exactly as if it had just arrived. That is the
 * whole reason this domain is thin: there is no second decryption path to keep in step with the
 * first.
 *
 * # Two things a caller owes
 *
 * Replay in **ascending sequence order**. A sender-key distribution has a lower sequence than the
 * content it unlocks, so replaying forward hands the messaging layer each key before the messages
 * that need it, and the pending buffer stays a fallback rather than the normal path.
 *
 * **De-duplicate.** Both crypto layers refuse a second decrypt of the same message, which is replay
 * protection working correctly; a caller that replayed an overlapping page twice would see history
 * that will not open.
 *
 * # The truncation boundary
 *
 * The server may cap how far back it serves. A [com.migo.core.protocol.SyncStatus.Truncated] response
 * means older history exists and was not returned, and a client should draw a visible boundary for
 * it. Rendering a truncated response as continuous is how a user comes to believe a conversation
 * started later than it did.
 */
class SyncDomain(private val rpc: Rpc) {
    /**
     * Fetches up to [limit] messages for a conversation.
     *
     * [haveSeq] is the highest contiguous sequence already held; the server returns what follows it,
     * or with [backwards] what precedes it. On success the caller replays [SyncResponse.messages] in
     * order and then advances its watermark to [SyncResponse.toSeq]; a [SyncResponse.more] of true
     * means another page is waiting, with that same `toSeq` as the next [haveSeq].
     *
     * [toSeq] bounds the far end, which is what fills a *detected gap* rather than tailing the
     * latest: a client that saw sequence 40 arrive while holding 30 asks for 31 to 39 and nothing
     * else, instead of re-fetching a page it will mostly discard.
     */
    suspend fun fetch(
        conversationId: Id,
        haveSeq: Long,
        limit: Long,
        toSeq: Long? = null,
        backwards: Boolean? = null,
    ): SyncResponse {
        val request = SyncRequest(conversationId, haveSeq, limit, toSeq, backwards)
        return rpc.call(Op.SYNC, { w -> request.encode(w) }, { r -> SyncResponse.decode(r) })
    }
}
